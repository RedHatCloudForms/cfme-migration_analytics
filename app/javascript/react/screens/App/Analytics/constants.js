export const PROVIDERS_SUMMARY_REPORT_FILTERS = {
  rpt_group: 'Migration Readiness - Providers',
  name: 'VMware Environment Summary'
};
