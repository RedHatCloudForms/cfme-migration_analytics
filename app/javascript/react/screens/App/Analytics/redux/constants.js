export const FETCH_MANIFEST_INFO = 'MA_FETCH_MANIFEST_INFO';
export const TOGGLE_MANIFEST_UPDATE_MODAL = 'MA_TOGGLE_MANIFEST_UPDATE_MODAL';
export const CALCULATE_SUMMARY_DATA = 'MA_CALCULATE_SUMMARY_DATA';
export const SELECT_PROVIDERS = 'MA_SELECT_PROVIDERS';
export const SELECT_DETAILED_DATA = 'MA_SELECT_DETAILED_DATA';
export const START_INVENTORY_BUNDLE = 'MA_START_INVENTORY_BUNDLE';
export const FETCH_BUNDLE_TASK = 'MA_FETCH_BUNDLE_TASK';
export const RESET_DATA_COLLECTION_STATE = 'MA_RESET_DATA_COLLECTION_STATE';

export const MANIFEST_INFO_URL = '/api/red_hat_migration_analytics';
export const INVENTORY_BUNDLE_URL = '/api/red_hat_migration_analytics';
