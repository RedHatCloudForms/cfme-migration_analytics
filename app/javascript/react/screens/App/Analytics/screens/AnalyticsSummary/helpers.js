export const someProvidersExist = ({ providers }) => providers && providers.length > 0;

export const noRefreshErrorsExist = ({ providersWithRefreshErrors }) =>
  providersWithRefreshErrors && providersWithRefreshErrors.length === 0;

export const someProvidersAwaitingRefresh = ({ providersAwaitingRefresh }) =>
  providersAwaitingRefresh && providersAwaitingRefresh.length > 0;

export const noProvidersAwaitingRefresh = ({ providersAwaitingRefresh }) =>
  providersAwaitingRefresh && providersAwaitingRefresh.length === 0;

export const providersRefreshed = props =>
  someProvidersExist(props) && noRefreshErrorsExist(props) && noProvidersAwaitingRefresh(props);
