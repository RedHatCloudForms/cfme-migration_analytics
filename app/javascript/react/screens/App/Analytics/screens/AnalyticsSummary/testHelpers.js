export const expectInitialFetchActions = (calledTimes, props) => {
  expect(props.fetchProvidersAction).toHaveBeenCalledTimes(calledTimes * 1);
  expect(props.fetchReportsAction).toHaveBeenCalledTimes(calledTimes * 2);
};

export const expectNoActionsOtherThan = (otherThanActions, props) => {
  const allActions = [
    'fetchProvidersAction',
    'fetchReportsAction',
    'runReportAction',
    'fetchTaskAction',
    'fetchResultAction',
    'calculateSummaryDataAction',
    'resetAllStateAction'
  ];
  allActions.forEach(action => {
    if (!otherThanActions.includes(action)) {
      const numCalls = props[action].mock.calls.length;
      if (numCalls > 0) {
        console.error(`${action} was called unexpectedly`); // eslint-disable-line no-console
      }
      expect(props[action]).toHaveBeenCalledTimes(0);
    }
  });
};

export const expectNoActions = props => expectNoActionsOtherThan([], props);

export const expectOnlyInitialFetchActions = (props, initialFetchTimes = 1) => {
  expectInitialFetchActions(initialFetchTimes, props);
  expectNoActionsOtherThan(['fetchProvidersAction', 'fetchReportsAction'], props);
};
