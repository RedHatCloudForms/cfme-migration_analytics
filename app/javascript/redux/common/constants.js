export const RESET_ALL_STATE = 'MA_RESET_ALL_STATE';
