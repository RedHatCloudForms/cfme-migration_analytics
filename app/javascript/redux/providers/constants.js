// Action types
export const FETCH_PROVIDERS = 'MA_FETCH_PROVIDERS';

// API URLs
export const PROVIDERS_URL = '/api/providers';
