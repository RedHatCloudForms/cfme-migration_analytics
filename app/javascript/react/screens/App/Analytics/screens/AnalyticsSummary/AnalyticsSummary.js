import React from 'react';
import PropTypes from 'prop-types';
import { Spinner } from 'patternfly-react';
import { VM_SUMMARY_REPORT_FILTERS, FINISHED, OK } from '../../constants';

// TODO sort out data we need from the mocks and write helpers to calculate it
// TODO build accordion showing summary data
// TODO provider selection?
// TODO next page?

class AnalyticsSummary extends React.Component {
  constructor(props) {
    super(props);
    this.taskFetchTimeout = null;
  }

  componentDidMount() {
    const { fetchReportsAction } = this.props;
    fetchReportsAction(VM_SUMMARY_REPORT_FILTERS); // Kicks off the chain of action calls in componentDidUpdate
  }

  componentDidUpdate(prevProps) {
    const {
      vmSummaryReport,
      runReportAction,
      vmSummaryReportRun,
      fetchTaskAction,
      isFetchingVmSummaryReportTask,
      vmSummaryReportTask,
      fetchResultAction,
      vmSummaryReportResult,
      calculateSummaryDataAction
    } = this.props;

    // Once we've found the report href, run it.
    if (!prevProps.vmSummaryReport && vmSummaryReport) {
      runReportAction(vmSummaryReport.href);
      return;
    }

    // Once we have a task href for the running report, fetch it.
    if (!prevProps.vmSummaryReportRun && vmSummaryReportRun) {
      fetchTaskAction(vmSummaryReportRun.task_href);
      return;
    }

    // If we fetched an unfinished task, wait and fetch it again.
    if (
      prevProps.isFetchingVmSummaryReportTask &&
      !isFetchingVmSummaryReportTask &&
      vmSummaryReportTask &&
      vmSummaryReportTask.state !== FINISHED
    ) {
      this.taskFetchTimeout = setTimeout(() => fetchTaskAction(vmSummaryReportRun.task_href), 1000);
      return;
    }

    // Once we have a successfully finished task, fetch its result.
    if (
      (!prevProps.vmSummaryReportTask || prevProps.vmSummaryReportTask.state !== FINISHED) &&
      vmSummaryReportTask &&
      vmSummaryReportTask.state === FINISHED &&
      vmSummaryReportTask.status === OK
    ) {
      clearTimeout(this.fetchTaskTimeout); // Just in case.
      fetchResultAction(vmSummaryReportRun.result_href);
      return;
    }

    // Once we have a report result, process it for rendering.
    if (!prevProps.vmSummaryReportResult && vmSummaryReportResult) {
      calculateSummaryDataAction(vmSummaryReportResult);
    }
  }

  render() {
    const { summaryData } = this.props;
    if (!summaryData) {
      return (
        <div className="large-spinner">
          <Spinner loading size="lg" inline />
          <h3>{__('Examining virtualization providers')}</h3>
        </div>
      );
    }

    // TODO handle case where vmSummaryReportTask has an error

    return (
      <React.Fragment>
        <h1>TODO</h1>
        <pre>{JSON.stringify(summaryData, 2)}</pre>
      </React.Fragment>
    );
  }
}

AnalyticsSummary.propTypes = {
  fetchReportsAction: PropTypes.func,
  vmSummaryReport: PropTypes.shape({
    href: PropTypes.string,
    name: PropTypes.string,
    rpt_group: PropTypes.string
  }),
  runReportAction: PropTypes.func,
  vmSummaryReportRun: PropTypes.shape({
    href: PropTypes.string,
    result_href: PropTypes.string,
    task_href: PropTypes.string
  }),
  fetchTaskAction: PropTypes.func,
  isFetchingVmSummaryReportTask: PropTypes.bool,
  vmSummaryReportTask: PropTypes.shape({
    id: PropTypes.string,
    state: PropTypes.string,
    status: PropTypes.string
  }),
  fetchResultAction: PropTypes.func,
  vmSummaryReportResult: PropTypes.shape({
    result_set: PropTypes.arrayOf(
      PropTypes.shape({
        allocated_disk_storage: PropTypes.number,
        mem_cpu: PropTypes.number,
        cpu_total_cores: PropTypes.number,
        'ext_management_system.name': PropTypes.string,
        'ext_management_system.id': PropTypes.number
      })
    )
  }),
  calculateSummaryDataAction: PropTypes.func,
  summaryData: PropTypes.shape({
    total: PropTypes.shape({
      numProviders: PropTypes.number,
      numHypervisors: PropTypes.number,
      numVms: PropTypes.number,
      allocatedDiskSpace: PropTypes.number,
      allocatedMemory: PropTypes.number,
      numCpuCores: PropTypes.number
    }),
    providers: PropTypes.arrayOf(
      PropTypes.shape({
        numHypervisors: PropTypes.number,
        numVms: PropTypes.number,
        allocatedDiskSpace: PropTypes.number,
        allocatedMemory: PropTypes.number,
        numCpuCores: PropTypes.number
      })
    )
  })
};

export default AnalyticsSummary;
