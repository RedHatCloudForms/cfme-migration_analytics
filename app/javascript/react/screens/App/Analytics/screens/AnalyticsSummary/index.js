import { connect } from 'react-redux';
import AnalyticsSummary from './AnalyticsSummary';
import { fetchReportsAction, runReportAction } from '../../../../../../redux/reports/reportsActions';
import { selectReportByFilterValues, selectRunByReport } from '../../../../../../redux/reports/reportsSelectors';
import { VM_SUMMARY_REPORT_FILTERS } from '../../constants';

const mapStateToProps = state => {
  const vmSummaryReport = selectReportByFilterValues(state, VM_SUMMARY_REPORT_FILTERS);
  return {
    vmSummaryReport,
    lastVmSummaryReportRun: selectRunByReport(state, vmSummaryReport)
  };
};

export default connect(
  mapStateToProps,
  { fetchReportsAction, runReportAction }
)(AnalyticsSummary);
