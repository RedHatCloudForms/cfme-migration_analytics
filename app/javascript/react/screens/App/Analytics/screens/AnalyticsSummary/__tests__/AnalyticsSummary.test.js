import React from 'react';
import { shallow } from 'enzyme';
import AnalyticsSummary from '../AnalyticsSummary';
import { ERROR, FINISHED, OK } from '../../../constants';
import {
  expectInitialFetchActions,
  expectOnlyInitialFetchActions,
  expectNoActions,
  expectNoActionsOtherThan
} from '../testHelpers';

jest.useFakeTimers();

describe('analytics summary screen', () => {
  const getBaseProps = () => ({
    fetchProvidersAction: jest.fn(),
    fetchReportsAction: jest.fn(),
    runReportAction: jest.fn(),
    fetchTaskAction: jest.fn(),
    fetchResultAction: jest.fn(),
    calculateSummaryDataAction: jest.fn(),
    onCollectInventoryClick: jest.fn(),
    resetAllStateAction: jest.fn()
  });

  const providersRefreshingProps = {
    providers: [{ mock: 'provider' }],
    providersWithRefreshErrors: [],
    providersAwaitingRefresh: [{ mock: 'provider' }]
  };

  const providersReadyProps = {
    providers: [{ mock: 'provider' }],
    providersWithRefreshErrors: [],
    providersAwaitingRefresh: []
  };

  const reportsReadyProps = {
    vmSummaryReport: { mock: 'report', href: 'mock vm report href' },
    envSummaryReport: { mock: 'report', href: 'mock env report href' }
  };

  const reportsRanProps = {
    vmSummaryReportRun: { mock: 'report run', task_href: 'mock vm task href', result_href: 'mock vm result href' },
    envSummaryReportRun: { mock: 'report run', task_href: 'mock env task href', result_href: 'mock env result href' }
  };

  const reportTasksFinishedProps = {
    vmSummaryReportTask: { mock: 'task', state: FINISHED, status: OK },
    envSummaryReportTask: { mock: 'task', state: FINISHED, status: OK }
  };

  const reportResultProps = {
    vmSummaryReportResult: { mock: 'result' },
    envSummaryReportResult: { mock: 'result' }
  };

  describe('renders correctly', () => {
    test('while running reports with no errors (default props)', () => {
      const component = shallow(<AnalyticsSummary {...getBaseProps()} />);
      expect(component).toMatchSnapshot();
    });

    test('when there is an error fetching providers', () => {
      const props = {
        ...getBaseProps(),
        errorFetchingProviders: { message: 'dummy error message' }
      };
      const component = shallow(<AnalyticsSummary {...props} />);
      expect(component).toMatchSnapshot();
    });

    test('when providers are missing', () => {
      const props = {
        ...getBaseProps(),
        providers: []
      };
      const component = shallow(<AnalyticsSummary {...props} />);
      expect(component).toMatchSnapshot();
    });

    test('when there are errors refreshing providers', () => {
      const props = {
        ...getBaseProps(),
        providersWithRefreshErrors: [{ mock: 'provider' }]
      };
      const component = shallow(<AnalyticsSummary {...props} />);
      expect(component).toMatchSnapshot();
    });

    test('when there is an error fetching reports', () => {
      const props = {
        ...getBaseProps(),
        errorFetchingReports: { message: 'dummy error message' }
      };
      const component = shallow(<AnalyticsSummary {...props} />);
      expect(component).toMatchSnapshot();
    });

    test('when there is an error running the VM report', () => {
      const props = {
        ...getBaseProps(),
        vmSummaryReportTask: { mock: 'task', status: ERROR }
      };
      const component = shallow(<AnalyticsSummary {...props} />);
      expect(component).toMatchSnapshot();
    });

    test('when there is an error running the environment report', () => {
      const props = {
        ...getBaseProps(),
        envSummaryReportTask: { mock: 'task', status: ERROR }
      };
      const component = shallow(<AnalyticsSummary {...props} />);
      expect(component).toMatchSnapshot();
    });

    test('when there are providers awaiting refresh', () => {
      const props = {
        ...getBaseProps(),
        providersAwaitingRefresh: [{ mock: 'provider' }]
      };
      const component = shallow(<AnalyticsSummary {...props} />);
      expect(component).toMatchSnapshot();
    });

    test('while fetching providers', () => {
      const props = {
        ...getBaseProps(),
        isFetchingProviders: true
      };
      const component = shallow(<AnalyticsSummary {...props} />);
      expect(component).toMatchSnapshot();
    });

    test('when summary data is ready', () => {
      const props = {
        ...getBaseProps(),
        summaryData: { mock: 'summary data' }
      };
      const component = shallow(<AnalyticsSummary {...props} />);
      expect(component).toMatchSnapshot();
    });
  });

  describe('initial fetch', () => {
    test('runs if providers are missing', () => {
      const props = getBaseProps();
      shallow(<AnalyticsSummary {...props} />);
      expectOnlyInitialFetchActions(props);
    });

    test('runs if VM summary report is missing', () => {
      const props = {
        ...getBaseProps(),
        providers: [{ mock: 'provider' }]
      };
      shallow(<AnalyticsSummary {...props} />);
      expectOnlyInitialFetchActions(props);
    });

    test('runs if environment summary report is missing', () => {
      const props = {
        ...getBaseProps(),
        providers: [{ mock: 'provider' }],
        vmSummaryReport: { mock: 'report' }
      };
      shallow(<AnalyticsSummary {...props} />);
      expectOnlyInitialFetchActions(props);
    });

    test('does not run if providers and both reports exist', () => {
      const props = {
        ...getBaseProps(),
        providers: [{ mock: 'provider' }],
        vmSummaryReport: { mock: 'report' },
        envSummaryReport: { mock: 'report' }
      };
      shallow(<AnalyticsSummary {...props} />);
      expectOnlyInitialFetchActions(props, 0);
    });

    test('resets all state and repeats initial fetch if retry button is clicked', () => {
      const props = getBaseProps();
      const component = shallow(<AnalyticsSummary {...props} />);
      expectOnlyInitialFetchActions(props);
      component.instance().retry();
      expect(props.resetAllStateAction).toHaveBeenCalledTimes(1);
      expectInitialFetchActions(2, props);
    });
  });

  describe('update lifecycle for providers', () => {
    test('calls no actions after provider fetch starts', () => {
      const props = getBaseProps();
      const component = shallow(<AnalyticsSummary {...props} />);
      expectOnlyInitialFetchActions(props);
      component.setProps({ isFetchingProviders: true });
      expectOnlyInitialFetchActions(props);
    });

    test('calls no actions after provider fetch comes back empty', () => {
      const props = {
        ...getBaseProps(),
        isFetchingProviders: true
      };
      const component = shallow(<AnalyticsSummary {...props} />);
      expectOnlyInitialFetchActions(props);
      component.setProps({ isFetchingProviders: false, providers: [] });
      expectOnlyInitialFetchActions(props);
    });

    test('calls no actions after provider fetch comes back with refresh errors', () => {
      const props = {
        ...getBaseProps(),
        isFetchingProviders: true
      };
      const component = shallow(<AnalyticsSummary {...props} />);
      expectOnlyInitialFetchActions(props);
      component.setProps({
        isFetchingProviders: false,
        providers: [{ mock: 'provider' }],
        providersWithRefreshErrors: [{ mock: 'provider' }]
      });
      expectOnlyInitialFetchActions(props);
    });

    test('polls providers once after provider fetch comes back with ongoing refresh', () => {
      const props = {
        ...getBaseProps(),
        isFetchingProviders: true
      };
      const component = shallow(<AnalyticsSummary {...props} />);
      expectOnlyInitialFetchActions(props);
      component.setProps({
        isFetchingProviders: false,
        ...providersRefreshingProps
      });
      expectOnlyInitialFetchActions(props);
      jest.runAllTimers();
      expect(props.fetchProvidersAction).toHaveBeenCalledTimes(2);
      expect(props.fetchReportsAction).toHaveBeenCalledTimes(2);
      expectNoActionsOtherThan(['fetchProvidersAction', 'fetchReportsAction'], props);
    });

    test('cancels polling if providers become ready before timer is up', () => {
      const props = {
        ...getBaseProps(),
        isFetchingProviders: true
      };
      const component = shallow(<AnalyticsSummary {...props} />);
      expectOnlyInitialFetchActions(props);
      component.setProps({
        isFetchingProviders: false,
        ...providersRefreshingProps
      });
      expect(component.instance().fetchProvidersTimeout).toBeTruthy();
      component.setProps({
        ...providersReadyProps
      });
      expect(component.instance().fetchProvidersTimeout).toBeFalsy();
      expectOnlyInitialFetchActions(props);
    });
  });

  describe('update lifecycle for reports', () => {
    test('calls no actions if reports were ready and nothing report-related changed', () => {
      const props = {
        ...getBaseProps(),
        ...providersReadyProps,
        vmSummaryReport: { mock: 'report' },
        envSummaryReport: { mock: 'report' }
      };
      const component = shallow(<AnalyticsSummary {...props} />);
      expectNoActions(props);
      component.setProps({ isFetchingProviders: true });
      expectNoActions(props);
    });

    test('runs VM report if it became ready', () => {
      const props = {
        ...getBaseProps(),
        ...providersReadyProps
      };
      const href = 'mock vm report href';
      const component = shallow(<AnalyticsSummary {...props} />);
      expectOnlyInitialFetchActions(props);
      component.setProps({
        vmSummaryReport: { mock: 'report', href }
      });
      expect(props.runReportAction).toHaveBeenCalledTimes(1);
      expect(props.runReportAction).toHaveBeenCalledWith(href);
      expectNoActionsOtherThan(['fetchProvidersAction', 'fetchReportsAction', 'runReportAction'], props);
    });

    test('runs environment report if it became ready', () => {
      const props = {
        ...getBaseProps(),
        ...providersReadyProps
      };
      const href = 'mock env report href';
      const component = shallow(<AnalyticsSummary {...props} />);
      expectOnlyInitialFetchActions(props);
      component.setProps({
        envSummaryReport: { mock: 'report', href }
      });
      expect(props.runReportAction).toHaveBeenCalledTimes(1);
      expect(props.runReportAction).toHaveBeenCalledWith(href);
      expectNoActionsOtherThan(['fetchProvidersAction', 'fetchReportsAction', 'runReportAction'], props);
    });

    test('runs both reports if they both became ready', () => {
      const props = {
        ...getBaseProps(),
        ...providersReadyProps
      };
      const vmHref = 'mock vm report href';
      const envHref = 'mock env report href';
      const component = shallow(<AnalyticsSummary {...props} />);
      expectOnlyInitialFetchActions(props);
      component.setProps({
        vmSummaryReport: { mock: 'report', href: vmHref },
        envSummaryReport: { mock: 'report', href: envHref }
      });
      expect(props.runReportAction).toHaveBeenCalledTimes(2);
      expect(props.runReportAction).toHaveBeenCalledWith(vmHref);
      expect(props.runReportAction).toHaveBeenCalledWith(envHref);
      expectNoActionsOtherThan(['fetchProvidersAction', 'fetchReportsAction', 'runReportAction'], props);
    });

    test('fetches task for VM report after it was just run', () => {
      const props = {
        ...getBaseProps(),
        ...providersReadyProps,
        ...reportsReadyProps
      };
      const taskHref = 'mock task href';
      const component = shallow(<AnalyticsSummary {...props} />);
      expectNoActions(props);
      component.setProps({
        vmSummaryReportRun: { mock: 'report run', task_href: taskHref }
      });
      expect(props.fetchTaskAction).toHaveBeenCalledTimes(1);
      expect(props.fetchTaskAction).toHaveBeenCalledWith(taskHref);
      expectNoActionsOtherThan(['fetchTaskAction'], props);
    });

    test('fetches task for environment report after it was just run', () => {
      const props = {
        ...getBaseProps(),
        ...providersReadyProps,
        ...reportsReadyProps
      };
      const taskHref = 'mock task href';
      const component = shallow(<AnalyticsSummary {...props} />);
      expectNoActions(props);
      component.setProps({
        envSummaryReportRun: { mock: 'report run', task_href: taskHref }
      });
      expect(props.fetchTaskAction).toHaveBeenCalledTimes(1);
      expect(props.fetchTaskAction).toHaveBeenCalledWith(taskHref);
      expectNoActionsOtherThan(['fetchTaskAction'], props);
    });

    test('fetches tasks for both reports if they both were just run', () => {
      const props = {
        ...getBaseProps(),
        ...providersReadyProps,
        ...reportsReadyProps
      };
      const vmTaskHref = 'mock vm task href';
      const envTaskHref = 'mock env task href';
      const component = shallow(<AnalyticsSummary {...props} />);
      expectNoActions(props);
      component.setProps({
        vmSummaryReportRun: { mock: 'report run', task_href: vmTaskHref },
        envSummaryReportRun: { mock: 'report run', task_href: envTaskHref }
      });
      expect(props.fetchTaskAction).toHaveBeenCalledTimes(2);
      expect(props.fetchTaskAction).toHaveBeenCalledWith(vmTaskHref);
      expect(props.fetchTaskAction).toHaveBeenCalledWith(envTaskHref);
      expectNoActionsOtherThan(['fetchTaskAction'], props);
    });

    test('polls VM report task once after it comes back unfinished', () => {
      const props = {
        ...getBaseProps(),
        ...providersReadyProps,
        ...reportsReadyProps,
        ...reportsRanProps,
        isFetchingVmSummaryReportTask: true
      };
      const component = shallow(<AnalyticsSummary {...props} />);
      expectNoActions(props);
      component.setProps({
        isFetchingVmSummaryReportTask: false,
        vmSummaryReportTask: { mock: 'task', state: 'Running' }
      });
      expectNoActions(props);
      jest.runAllTimers(props);
      expect(props.fetchTaskAction).toHaveBeenCalledTimes(1);
      expectNoActionsOtherThan(['fetchTaskAction'], props);
    });

    test('polls environment report task once after it comes back unfinished', () => {
      const props = {
        ...getBaseProps(),
        ...providersReadyProps,
        ...reportsReadyProps,
        ...reportsRanProps,
        isFetchingEnvSummaryReportTask: true
      };
      const component = shallow(<AnalyticsSummary {...props} />);
      expectNoActions(props);
      component.setProps({
        isFetchingEnvSummaryReportTask: false,
        envSummaryReportTask: { mock: 'task', state: 'Running' }
      });
      expectNoActions(props);
      jest.runAllTimers(props);
      expect(props.fetchTaskAction).toHaveBeenCalledTimes(1);
      expectNoActionsOtherThan(['fetchTaskAction'], props);
    });

    test('polls both report tasks once each if they both came back unfinished', () => {
      const props = {
        ...getBaseProps(),
        ...providersReadyProps,
        ...reportsReadyProps,
        ...reportsRanProps,
        isFetchingVmSummaryReportTask: true,
        isFetchingEnvSummaryReportTask: true
      };
      const component = shallow(<AnalyticsSummary {...props} />);
      expectNoActions(props);
      component.setProps({
        isFetchingVmSummaryReportTask: false,
        isFetchingEnvSummaryReportTask: false,
        vmSummaryReportTask: { mock: 'task', state: 'Running' },
        envSummaryReportTask: { mock: 'task', state: 'Running' }
      });
      expectNoActions(props);
      jest.runAllTimers(props);
      expect(props.fetchTaskAction).toHaveBeenCalledTimes(2);
      expectNoActionsOtherThan(['fetchTaskAction'], props);
    });

    test('cancels VM report task polling if task came back finished before timer is up', () => {
      const props = {
        ...getBaseProps(),
        ...providersReadyProps,
        ...reportsReadyProps,
        ...reportsRanProps,
        isFetchingVmSummaryReportTask: true
      };
      const component = shallow(<AnalyticsSummary {...props} />);
      expectNoActions(props);
      expect(component.instance().fetchTaskTimeouts).toEqual({});
      component.setProps({
        isFetchingVmSummaryReportTask: false,
        vmSummaryReportTask: { mock: 'task', state: 'Running' }
      });
      expect(Object.values(component.instance().fetchTaskTimeouts)[0]).toBeTruthy();
      expectNoActions(props);
      component.setProps({
        vmSummaryReportTask: { mock: 'task', state: FINISHED, status: OK }
      });
      expect(Object.values(component.instance().fetchTaskTimeouts)).toEqual([null]);
      expect(props.fetchResultAction).toHaveBeenCalledTimes(1);
      expectNoActionsOtherThan(['fetchResultAction'], props);
    });

    test('cancels environment report task polling if task came back finished before timer is up', () => {
      const props = {
        ...getBaseProps(),
        ...providersReadyProps,
        ...reportsReadyProps,
        ...reportsRanProps,
        isFetchingEnvSummaryReportTask: true
      };
      const component = shallow(<AnalyticsSummary {...props} />);
      expectNoActions(props);
      expect(component.instance().fetchTaskTimeouts).toEqual({});
      component.setProps({
        isFetchingEnvSummaryReportTask: false,
        envSummaryReportTask: { mock: 'task', state: 'Running' }
      });
      expect(Object.values(component.instance().fetchTaskTimeouts)[0]).toBeTruthy();
      expectNoActions(props);
      component.setProps({
        envSummaryReportTask: { mock: 'task', state: FINISHED, status: OK }
      });
      expect(Object.values(component.instance().fetchTaskTimeouts)).toEqual([null]);
      expect(props.fetchResultAction).toHaveBeenCalledTimes(1);
      expectNoActionsOtherThan(['fetchResultAction'], props);
    });

    test('cancels both report tasks polling if both tasks came back finished before timers were up', () => {
      const props = {
        ...getBaseProps(),
        ...providersReadyProps,
        ...reportsReadyProps,
        ...reportsRanProps,
        isFetchingVmSummaryReportTask: true,
        isFetchingEnvSummaryReportTask: true
      };
      const component = shallow(<AnalyticsSummary {...props} />);
      expectNoActions(props);
      expect(component.instance().fetchTaskTimeouts).toEqual({});
      component.setProps({
        isFetchingVmSummaryReportTask: false,
        isFetchingEnvSummaryReportTask: false,
        vmSummaryReportTask: { mock: 'task', state: 'Running' },
        envSummaryReportTask: { mock: 'task', state: 'Running' }
      });
      const timeouts = Object.values(component.instance().fetchTaskTimeouts);
      expect(timeouts).toHaveLength(2);
      timeouts.map(timeout => expect(timeout).toBeTruthy);
      expectNoActions(props);
      component.setProps({
        ...reportTasksFinishedProps
      });
      expect(Object.values(component.instance().fetchTaskTimeouts)).toEqual([null, null]);
      expect(props.fetchResultAction).toHaveBeenCalledTimes(2);
      expectNoActionsOtherThan(['fetchResultAction'], props);
    });

    test('fetches results if VM report task came back finished', () => {
      const props = {
        ...getBaseProps(),
        ...providersReadyProps,
        ...reportsReadyProps,
        ...reportsRanProps,
        isFetchingVmSummaryReportTask: true
      };
      const component = shallow(<AnalyticsSummary {...props} />);
      expectNoActions(props);
      component.setProps({
        isFetchingVmSummaryReportTask: false,
        isFetchingEnvSummaryReportTask: false,
        vmSummaryReportTask: { mock: 'task', state: FINISHED, status: OK }
      });
      expect(props.fetchResultAction).toHaveBeenCalledTimes(1);
      expect(props.fetchResultAction).toHaveBeenCalledWith('mock vm result href');
      expectNoActionsOtherThan(['fetchResultAction'], props);
    });

    test('fetches results if environment report task came back finished', () => {
      const props = {
        ...getBaseProps(),
        ...providersReadyProps,
        ...reportsReadyProps,
        ...reportsRanProps,
        isFetchingEnvSummaryReportTask: true
      };
      const component = shallow(<AnalyticsSummary {...props} />);
      expectNoActions(props);
      component.setProps({
        isFetchingVmSummaryReportTask: false,
        isFetchingEnvSummaryReportTask: false,
        envSummaryReportTask: { mock: 'task', state: FINISHED, status: OK }
      });
      expect(props.fetchResultAction).toHaveBeenCalledTimes(1);
      expect(props.fetchResultAction).toHaveBeenCalledWith('mock env result href');
      expectNoActionsOtherThan(['fetchResultAction'], props);
    });

    test('fetches both results if both reports came back finished', () => {
      const props = {
        ...getBaseProps(),
        ...providersReadyProps,
        ...reportsReadyProps,
        ...reportsRanProps,
        isFetchingVmSummaryReportTask: true,
        isFetchingEnvSummaryReportTask: true
      };
      const component = shallow(<AnalyticsSummary {...props} />);
      expectNoActions(props);
      component.setProps({
        isFetchingVmSummaryReportTask: false,
        isFetchingEnvSummaryReportTask: false,
        ...reportTasksFinishedProps
      });
      expect(props.fetchResultAction).toHaveBeenCalledTimes(2);
      expect(props.fetchResultAction).toHaveBeenCalledWith('mock vm result href');
      expect(props.fetchResultAction).toHaveBeenCalledWith('mock env result href');
      expectNoActionsOtherThan(['fetchResultAction'], props);
    });

    test('calculates summary data if all necessary report results became ready', () => {
      const props = {
        ...getBaseProps(),
        ...providersReadyProps,
        ...reportsReadyProps,
        ...reportsRanProps,
        ...reportTasksFinishedProps
      };
      const component = shallow(<AnalyticsSummary {...props} />);
      expectNoActions(props);
      component.setProps({
        ...reportResultProps
      });
      expect(props.calculateSummaryDataAction).toHaveBeenCalledTimes(1);
      expect(props.calculateSummaryDataAction).toHaveBeenCalledWith(reportResultProps);
      expectNoActionsOtherThan(['calculateSummaryDataAction'], props);
    });

    test('does not recalculate summary data if all report results were already ready on update', () => {
      const props = {
        ...getBaseProps(),
        ...providersReadyProps,
        ...reportsReadyProps,
        ...reportsRanProps,
        ...reportTasksFinishedProps,
        ...reportResultProps
      };
      const component = shallow(<AnalyticsSummary {...props} />);
      expectNoActions(props);
      component.setProps({
        isFetchingProviders: true
      });
      expectNoActions(props);
    });
  });
});
