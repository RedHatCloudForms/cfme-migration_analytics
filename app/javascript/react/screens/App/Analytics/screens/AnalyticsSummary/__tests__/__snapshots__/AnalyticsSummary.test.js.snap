// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`analytics summary screen renders correctly when providers are missing 1`] = `
<EmptyStateWithButton
  buttonText="Configure Infrastructure Providers"
  href="/ems_infra/show_list"
  iconName="add-circle-o"
  iconType="pf"
  message="Before collecting analytics data, you must have at least one VMware provider configured."
  onClick={[Function]}
  title="Missing Providers"
/>
`;

exports[`analytics summary screen renders correctly when summary data is ready 1`] = `
<Fragment>
  <SummaryAccordion
    summaryData={
      Object {
        "mock": "summary data",
      }
    }
  />
  <Button
    active={false}
    block={false}
    bsClass="btn"
    bsStyle="primary"
    disabled={false}
    onClick={[MockFunction]}
  >
    Collect Inventory Data
  </Button>
</Fragment>
`;

exports[`analytics summary screen renders correctly when there are errors refreshing providers 1`] = `
<ProviderRefreshErrors
  onRetryClick={[Function]}
  providersWithRefreshErrors={
    Array [
      Object {
        "mock": "provider",
      },
    ]
  }
/>
`;

exports[`analytics summary screen renders correctly when there are providers awaiting refresh 1`] = `
<LargeInlineSpinner
  message="Waiting for provider refresh"
/>
`;

exports[`analytics summary screen renders correctly when there is an error fetching providers 1`] = `
<EmptyStateWithButton
  buttonText="Try again"
  href={null}
  iconName="error-circle-o"
  iconType="pf"
  message="dummy error message"
  onClick={[Function]}
  title="Failed to fetch providers"
/>
`;

exports[`analytics summary screen renders correctly when there is an error fetching reports 1`] = `
<EmptyStateWithButton
  buttonText="Try again"
  href={null}
  iconName="error-circle-o"
  iconType="pf"
  message="dummy error message"
  onClick={[Function]}
  title="Failed to fetch reports"
/>
`;

exports[`analytics summary screen renders correctly when there is an error running the VM report 1`] = `
<ReportTaskError
  onRetryClick={[Function]}
  taskWithError={
    Object {
      "mock": "task",
      "status": "Error",
    }
  }
/>
`;

exports[`analytics summary screen renders correctly when there is an error running the environment report 1`] = `
<ReportTaskError
  onRetryClick={[Function]}
  taskWithError={
    Object {
      "mock": "task",
      "status": "Error",
    }
  }
/>
`;

exports[`analytics summary screen renders correctly while fetching providers 1`] = `
<LargeInlineSpinner
  message="Checking providers"
/>
`;

exports[`analytics summary screen renders correctly while running reports with no errors (default props) 1`] = `
<LargeInlineSpinner
  message="Examining virtualization providers"
/>
`;
